### PR TITLE
DPI feature

### PR DESCRIPTION
**Pull request by garyachy**
Implementation for CLI commands:
* create upf application <name>
* delete upf application <name>
* show upf applications
* upf application <name> rule <id> <rule>
* upf application <name> rule <id> del
* show upf application <name>

### CODE REVIEW

File: src/plugins/upf/upf.c [hunk 1]
```diff
@@ -773,6 +773,236 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+  pool_get (sm->dpi_apps_pool, app);
+  memset(app, 0, sizeof(*app));
+
+  /* Get a line of input. */
+  if (unformat_user (input, unformat_line_input, line_input))
+  {
+    while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "%s", &name))
+      {
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: vpp uses GNU code style, this line and other are not indented in GNU style

File: src/plugins/upf/upf.c [hunk 2]
```diff
@@ -773,6 +773,236 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+        app->name = vec_dup(name);
+        hash_set_mem (sm->dpi_app_hash, app->name, app - sm->dpi_apps_pool);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: It is common practice in VPP to first parse the full line and the execute the action.

The logic here would prine an error and still add a entry for an input line that start with valid input and end with garbage.

Please, have a look the other CLI parsers in this file and mirror their style.

File: src/plugins/upf/upf.c [hunk 3]
```diff
@@ -773,6 +773,236 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+    {
+      if (unformat (line_input, "%s rule %u %s",
+                    &app_name, &rule_index, &rule_name))
+      {
+        index = hash_get_mem (sm->dpi_app_hash, app_name);
+        if (index)
+        {
+          vlib_cli_output (vm, "Application %s is present", app_name);
+
+          if (strcmp ((char*)rule_name, "del") == 0)
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: should use unformat

File: src/plugins/upf/upf.c [hunk 4]
```diff
@@ -773,6 +773,236 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+VLIB_CLI_COMMAND (upf_delete_app_command, static) =
+{
+  .path = "delete upf application",
+  .short_help = "delete upf application <name>",
+  .function = upf_delete_app_command_fn,
+};
+/* *INDENT-ON* */
+
+static clib_error_t *
+upf_create_delete_rule_command_fn (vlib_main_t * vm,
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: a more comon style in VPP would be to name this ```upf_application_rule_add_del_command_fn```

File: src/plugins/upf/upf.h [hunk 5]
```diff
@@ -490,6 +502,11 @@ typedef struct {
   vlib_main_t * vlib_main;
   vnet_main_t * vnet_main;
   ethernet_main_t * ethernet_main;
+
+  /* DPI apps hash */
+  uword* dpi_app_hash;
+  /* DPI apps vector */
+  upf_dpi_app_t *dpi_apps_pool;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: pls use ```upf_apps``` instead of dpi_apps_pool, the first use case will be dpi, but this will also be used for other stuff

File: src/plugins/upf/upf.h [hunk 6]
```diff
@@ -490,6 +502,11 @@ typedef struct {
   vlib_main_t * vlib_main;
   vnet_main_t * vnet_main;
   ethernet_main_t * ethernet_main;
+
+  /* DPI apps hash */
+  uword* dpi_app_hash;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: ```upf_app_by_name```

File: src/plugins/upf/upf.h [hunk 7]
```diff
@@ -437,6 +437,18 @@ typedef struct {
...
 } upf_node_assoc_t;
 
+typedef struct {
+  u32 id;
+} upf_dpi_rule_t;
+
+typedef struct {
+  u8 * name;
+  /* Rules hash */
+  uword* rules_hash;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: ```rules_by_id```

File: src/plugins/upf/upf.h [hunk 8]
```diff
@@ -437,6 +437,18 @@ typedef struct {
...
+  /* Rules vector */
+  upf_dpi_rule_t *rules_pool;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: simply ```rules```

File: src/plugins/upf/upf.c [hunk 9]
```diff
@@ -773,6 +773,263 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
 };
 /* *INDENT-ON* */
 
+static clib_error_t *
+upf_create_app_command_fn (vlib_main_t * vm,
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: don't reinvent the wheel. The nwi CLI functions do something very similar.

Use the upf_nwi_add_del_command_fn and the vnet_upf_nwi_add_del functions as blueprint for this.

File: src/plugins/upf/upf.c [hunk 10]
```diff
@@ -773,6 +773,380 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+upf_application_rule_add_del_command_fn (vlib_main_t * vm,
+                                         unformat_input_t * input,
+                                         vlib_cli_command_t * cmd)
+{
+  unformat_input_t _line_input, *line_input = &_line_input;
+  u8 *app_name = NULL;
+  u32 rule_index = 0;
+  clib_error_t *error = NULL;
+  int rv = 0;
+  int add = 0;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: this should be initialized to 1

File: src/plugins/upf/upf.c [hunk 11]
```diff
@@ -773,6 +773,380 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+      if (unformat (line_input, "%s rule %u",
+                    &app_name, &rule_index))
+        {
+          if (unformat (line_input, "del"))
+            {
+              add = 0;
+              break;
+            }
+          else
+            {
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: either loose the ```else``` block or explicitly test for an ```add``` like here https://github.com/travelping/vpp/blob/feature/upf/src/plugins/upf/upf.c#L210

File: src/plugins/upf/upf.c [hunk 12]
```diff
@@ -773,6 +799,473 @@ VLIB_CLI_COMMAND (upf_show_session_command, static) =
...
+          if (unformat (line_input, "del"))
+            {
+              add = 0;
+              break;
+            }
+          else if (unformat (line_input, "add"))
+            {
+              add = 1;
+
+              if (unformat (line_input, "ip dst %s", &dst_ip))
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: ```unformat (line_input, "%U", unformat_ip46_address, &ip, IP46_TYPE_ANY)``` would parse a IP address to binary form. No need to pass strings around.

File: src/plugins/upf/upf_pfcp_api.c [hunk 13]
```diff
@@ -855,6 +856,21 @@ static int handle_create_pdr(upf_session_t *sess, pfcp_create_pdr_t *create_pdr,
...
+
+	if (ISSET_BIT(pdr->pdi.grp.fields, PDI_APPLICATION_ID))
+		{
+			create->pdi.fields |= F_PDI_APPLICATION_ID;
+
+			pfcp_application_id_t *app_id;
+	
+			vec_foreach(app_id, pdr->pdi.application_id)
+				{
+					vec_add1(create->pdi.app_id, *app_id);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_pfcp_api.c]: vec_add1 almost immediately followed by a vec_free feels wrong. Something must be wrong here....

File: src/plugins/upf/upf_classify.c [hunk 14]
```diff
@@ -170,7 +200,14 @@ upf_classify (vlib_main_t * vm, vlib_node_runtime_t * node,
 		    {
 		      vnet_buffer (b)->gtpu.pdr_idx = pdr - active->pdr;
 		      far = sx_get_far_by_id(active, pdr->far_id);
+
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_classify.c]: Most likely is this not the right place to integrate the DPI scanning into the classify path....

File: src/plugins/upf/upf_pfcp_api.c (hunk 13 above)
Comment (yuriyklymenko) [reply src/plugins/upf/upf_pfcp_api.c]: My idea is to use a temporary vector as a parameter. The vector contains a list of app ids.

File: src/plugins/upf/dpi.c [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+upf_dpi_lookup(u32 db_index, u8 * str, uint16_t length, u32 * app_index)
+{
+  upf_dpi_entry_t *entry = NULL;
+  int ret = 0;
+  upf_dpi_cb_args_t args = {};
+
+  if (!db_index)
+    return -1;
+
+  if (pool_elts(upf_dpi_db) < db_index)
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: This is wrong for two reasons:
1. pool_elts returns the number of active elements, no the index of highest index that is used in the pool. It is therefore possible that pool_elts  < db_index for a valid db_index.
2. the pool_elt_at_index is testing the index anyways, both for being greater than the maximum index and for not being free.

File: src/plugins/upf/dpi.c [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+int
+upf_dpi_lookup(u32 db_index, u8 * str, uint16_t length, u32 * app_index)
+{
+  upf_dpi_entry_t *entry = NULL;
+  int ret = 0;
+  upf_dpi_cb_args_t args = {};
+
+  if (!db_index)
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: VPP is not completely consistent on the use of index zero as valid or invalid. However the UPF plugin usually accepts a zero index as **valid** and uses ```~0``` as the invalid value.

I would prefer is this could be done for the db_index as well.

File: src/plugins/upf/upf.c [hunk 17]
```diff
@@ -31,8 +31,318 @@
...
+#include <upf/dpi.h>
+
+static void
+foreach_upf_flows (BVT (clib_bihash_kv) * kvp, void * arg);
+
+static void
+upf_add_rules(u32 app_index, upf_dpi_app_t *app, upf_dpi_args_t ** args)
+{
+  u32 index = 0;
+  u32 rule_index = 0;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.c]: it is not necessary to initialize the variable. hash_foreach will do that.

File: src/plugins/upf/upf_classify.c (hunk 14 above)
Comment (RoadRunnr) [reply src/plugins/upf/upf_classify.c]: Even after some changes, this still feels wrong.

If we have already detected a flow, there should a point in the life of the flow where we know which PDR, FAR and URR applies to it. Redoing all the PDR and FAR matching and lookup should be skipped at that point in time. I'm missing the logic that handles this.

**Comment by RoadRunnr** [comment]
At this point, there are still lots of whitespace errors, i.e. spaces instead of tabs, trailing whitespace

File: src/plugins/upf/pfcp.h [hunk 18]
```diff
@@ -190,7 +190,7 @@ typedef struct {
 } pfcp_sdf_filter_t;
 
 #define PFCP_IE_APPLICATION_ID				24
-typedef u32 pfcp_application_id_t;
+typedef u8 * pfcp_application_id_t;
```
Comment (RoadRunnr) [inline src/plugins/upf/pfcp.h]: The feature/upf was update to implement this (and your version is also wrong). This needs to be rebased to the current head of feature/upf

File: src/plugins/upf/upf_pfcp_api.c [hunk 19]
```diff
@@ -855,6 +856,23 @@ static int handle_create_pdr(upf_session_t *sess, pfcp_create_pdr_t *create_pdr,
...
 	}
+
+	if (ISSET_BIT(pdr->pdi.grp.fields, PDI_APPLICATION_ID))
+		{
+			create->pdi.fields |= F_PDI_APPLICATION_ID;
+
+			pfcp_application_id_t *app_id = NULL;
+			pfcp_application_id_t *app_id_vector = NULL;
+	
+			vec_foreach(app_id, pdr->pdi.application_id)
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_pfcp_api.c]: The PDI Application Id field is not an array, there can only be one App Id in a PDI (see Table 7.5.2.2-2 in 3GPP TS 29.244)

File: src/plugins/upf/upf_pfcp_api.c [hunk 20]
```diff
@@ -871,14 +865,8 @@ static int handle_update_pdr(upf_session_t *sess, pfcp_update_pdr_t *update_pdr,
...
-				pfcp_application_id_t *app_id = NULL;
 				pfcp_application_id_t *app_id_vector = NULL;
-		
-				vec_foreach(app_id, pdr->pdi.application_id)
-					{
-						vec_add1(app_id_vector, *app_id);
-					}
-
+				vec_add1(app_id_vector, pdr->pdi.application_id);
 				upf_add_multi_regex(app_id_vector, &update->dpi_db_id, 0);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_pfcp_api.c]: since there is now only one Application Id to add, the whole multi_regex logic could be simplified. The only called that can add multiple App Id is the CLI, and that can be changed.

File: src/plugins/upf/dpi.c [hunk 21]
```diff
@@ -0,0 +1,1052 @@
...
+    return -1;
+
+  *expressions = entry->expressions;
+  *ids = entry->ids;
+
+  return 0;
+}
+
+int
+upf_dpi_add_multi_regex(upf_dpi_args_t * args, u32 * db_index, u8 create)
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: Could you elaborate why you are using an entry in the dpi_db per activated PDI ?

To me it seems that you are compiling a regex for every PDI that gets activate. Since we don't need to merge different applications into a big regex, that seems to be a lot overhead. Compiling the regex per application only when the application is initially loaded should suffice.

I would understand if you merged all DPI rules for a session into a big scan expression and the did some reprocessing to find out if a match was a false positive. But that seems not to be the case.

**Comment by garyachy** [comment]
Andreas, it was done that way, because initially, we all thought that one PDR can be associated with multiple apps. Now, I agree that this logic can be simplified. 

My concern is that the current solution can support one or many apps. And a simplified one will support one app only. But it is not a big concern as far as the code is simple enough to change in the future.

File: src/plugins/upf/upf_classify.c (hunk 14 above)
Comment (RoadRunnr) [reply src/plugins/upf/upf_classify.c]: This has not been resolved. The flow tracking is still attached in the wrong manner.

File: src/plugins/upf/upf_pfcp_api.c [hunk 22]
```diff
@@ -737,6 +740,19 @@ static int handle_create_pdr(upf_session_t *sess, pfcp_create_pdr_t *create_pdr,
 	      break;
 	    }
 	}
+
+	if (ISSET_BIT(pdr->pdi.grp.fields, PDI_APPLICATION_ID))
+		{
+			create->pdi.fields |= F_PDI_APPLICATION_ID;
+
+			create->app_name = vec_dup(pdr->pdi.application_id);
+			vec_add1(create->app_name, 0);
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_pfcp_api.c]: This should not be a zero terminated string.

File: src/plugins/upf/dpi.c [hunk 23]
```diff
@@ -0,0 +1,1123 @@
...
+  clib_error_t *error = NULL;
+  int rv = 0;
+
+  /* Get a line of input. */
+  if (!unformat_user (input, unformat_line_input, line_input))
+    return error;
+
+  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "%s", &name))
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: ```unformat (line_input, "%v", &name)``` would convert it into a vector without the zero termination (see also https://github.com/travelping/vpp/pull/5/files#r220119212)

Comment (RoadRunnr) [reply src/plugins/upf/dpi.c]: Update, VPP uses %v to scan strings into vectors (without the terminating 0), it uses %s to scan into the typical zero terminated C strings.
Please replace all place where it make sense where you use %s in unformat with %v.

File: src/plugins/upf/dpi.c [hunk 24]
```diff
@@ -0,0 +1,1123 @@
...
+  hs_free_scratch(entry->scratch);
+  vec_free(entry->expressions);
+  vec_free(entry->flags);
+  vec_free(entry->ids);
+
+  memset(entry, 0, sizeof(upf_dpi_entry_t));
+}
+
+int
+upf_dpi_get_db_contents(u32 db_index, regex_t ** expressions, u32 ** ids)
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: This method is used in only one place. Using ```pool_elt_at_index``` directly (without hiding it in the wrapper) would make the code more readable.

File: src/plugins/upf/dpi.h [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+  if (host == NULL)
+    return -1;
+
+  host_end = (u8*)strchr((const char*)host, '\r');
+  if (host_end == NULL)
+    return -1;
+
+  host_length = host_end - host;
+
+  res = upf_dpi_lookup(host_db_id, host,
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.h]: I'm really confused about the split into host and path db.

The way I read this, it can confuse host and path.

For example, lets assume we have this rule base:

```
upf application GOLD rule 1 add l7 http host host1 path /path1
upf application GOLD rule 2 add l7 http host host2 path /path2
```

Then the split logic would match ```http://host2/path1```. Clearly that would be wrong.

File: src/plugins/upf/upf_classify.c [hunk 26]
```diff
@@ -142,27 +152,69 @@ upf_classify (vlib_main_t * vm, vlib_node_runtime_t * node,
...
+		  pdr = sx_get_pdr_by_id(active, flow->server_pdr_id);
+		}
+	    }
+
+	  if (pdr == NULL)
+	    {
+	      if ((flow->client_direction != direction) && flow->app_index != ~0)
+	      {
+		pdr = upf_get_dpi_pdr_by_name(active, direction, flow->app_index);
+		flow->server_pdr_id = pdr->id;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf_classify.c]: pdr might be NULL, a check is needed

File: src/plugins/upf/upf.h [hunk 27]
```diff
@@ -259,6 +262,10 @@ typedef struct {
   u8 outer_header_removal;
   u16 far_id;
   u16 *urr_ids;
+  u8 *app_name;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: I'm not sure that app_name is really needed in here. In most places it is used only for output and those places could lookup the string from the db.

File: src/plugins/upf/dpi.h (hunk 25 above)
Comment (RoadRunnr) [reply src/plugins/upf/dpi.h]: The split db approach is IMHO not going to work even with the latest changes.

File: src/plugins/upf/flowtable.h [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+    flow_data_t infos;
+
+    /* L7 app index */
+    u32 app_index;
+
+    /* TCP client direction */
+    u8 client_direction;
+
+    /* Client PDR */
+    u32 client_pdr_id;
```
Comment (RoadRunnr) [inline src/plugins/upf/flowtable.h]: this should be called ```initiator``` instead of ```client```

File: src/plugins/upf/flowtable.h [hunk 29]
```diff
@@ -0,0 +1,212 @@
...
+
+    /* Server PDR */
+    u32 server_pdr_id;
```
Comment (RoadRunnr) [inline src/plugins/upf/flowtable.h]: ```responder``` instead of ```server```

File: src/plugins/upf/flowtable.h [hunk 30]
```diff
@@ -0,0 +1,212 @@
...
+    u32 pkts;
+    u64 Bytes;
+} flow_stats_t;
+
+typedef struct {
+    u32 flags;
+    u32 ctx_id;
+    u32 clt_pkts;
+    u32 srv_pkts;
+    u64 clt_Bytes;
```
Comment (RoadRunnr) [inline src/plugins/upf/flowtable.h]: lower case

File: src/plugins/upf.am [hunk 31]
```diff
@@ -54,7 +57,15 @@ upf_plugin_la_SOURCES = \
...
   upf/upf_pfcp_api.h \
   upf/upf_http_redirect_server.h \
-  upf/upf_http_redirect_server.c
+  upf/upf_http_redirect_server.c \
+  upf/flowdata.h \
+  upf/flowtable.h \
+  upf/flowtable_impl.c \
+  upf/flowtable_impl.h \
+  upf/flowtable_tcp.h \
+  upf/dpi.c \
```
Comment (RoadRunnr) [inline src/plugins/upf.am]: please rename this to ```upf_adf``` and replace dpi with adf everywhere.

ADF means "Application Detection Function"

File: src/plugins/upf/upf.h [hunk 32]
```diff
@@ -493,6 +503,33 @@ typedef struct {
   u32 sessions;
 } upf_node_assoc_t;
 
+typedef u8 * regex_t;
+
+typedef struct {
+  u32 id;
+  regex_t host;
+  regex_t path;
+} upf_dpi_rule_t;
```
Comment (RoadRunnr) [inline src/plugins/upf/upf.h]: shoud be ```upf_adr_t```

ADR = Application Detection Rule

File: src/plugins/upf/dpi.c [hunk 33]
```diff
@@ -0,0 +1,1133 @@
...
+  app = pool_elt_at_index(sm->upf_apps, p[0]);
+
+  *path_db_index = app->path_db_index; 
+  *host_db_index = app->host_db_index; 
+
+  return 0;
+}
+
+static int
+upf_dpi_create_update_db(u8 * app_name, u32 * path_db_index,
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: this whole function is strange.

There are two stages that are almost identical, they first create a vector of args, then iterate over that vector to create yet another vector that is then feed into hs_compile.
From my understanding of that code, build the intermediate vector can be avoided.

File: src/plugins/upf/dpi.c [hunk 34]
```diff
@@ -0,0 +1,1133 @@
...
+
+  res = upf_dpi_add_multi_regex(args, host_db_index);
+
+  vec_free(args);
+
+  return res;
+}
+
+static void
+upf_dpi_all_pdr_update(u8* app_name)
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.c]: why is that needed?

Does dpi_{path,host}_db_id change when rules are added? If so, why?

File: src/plugins/upf/dpi.h [hunk 35]
```diff
@@ -0,0 +1,220 @@
...
+  http += sizeof("GET");
+  tcp_payload_len -= sizeof("GET");
+
+  version = (u8*)strchr((const char*)http, ' ');
+  if (version == NULL)
+    return -1;
+
+  uri_length = version - http;
+
+  res = upf_dpi_lookup(path_db_id, http,
```
Comment (RoadRunnr) [inline src/plugins/upf/dpi.h]: Instead of doing two lookups, would it possible to regex that scans for packet for the Path and the Host in the same step?

Or maybe construct a URI from the packet and use that for the lookup?

**Comment by RoadRunnr** [comment]
Something is strange. With this PR applied, the startup time of VPP increases dramatically.

**Comment by garyachy** [comment]
I do not see any delay in the latest code.
Please describe how much time it takes from your side and how you measure
it.
